libbpf-rs/src/btf/types.rs

```diff
@@ -863,26 +844,333 @@ gen_collection_concrete_type! {
+     * This is similar to simple-match but it's hardcoded for composite which matches both structs
+     * and unions.
+     */
+    (
+        match $ty:ident.kind() { $($p:pat => $a:expr),* }
+        BtfKind::Composite $( ($var:ident) )? => $action:expr,
+        $($rest:tt)*
+    ) => {
+        $crate::__btf_type_match!(match $ty.kind() { $($p => $a,)* }
+            BtfKind::Composite $( ($var) )* => { $action }
+            $($rest)*
+        )
+    };
+    (
+        match $ty:ident.kind() { $($p:pat => $a:expr),* }
+        BtfKind::Composite $(($var:ident))? => $action:block
+        $($rest:tt)*
+    ) => {
+        $crate::__btf_type_match!(match $ty.kind() {
+            $($p => $a,)*
+            $crate::btf::BtfKind::Struct | $crate::btf::BtfKind::Union => {
+                $(let $var = $crate::btf::types::Composite::try_from($ty).unwrap();)*
+                $action
+            }
         }
-    }}
+             $($rest)*
+        )
+    };
+    // simple-match: match on simple patterns that use an expression followed by a comma
+    (
+        match $ty:ident.kind() { $($p:pat => $a:expr),* }
+        BtfKind::$name:ident $(($var:ident))? => $action:expr,
+        $($rest:tt)*
+    ) => {
+        $crate::__btf_type_match!(
+            match $ty.kind() { $($p => $a),* }
+            BtfKind::$name $(($var))? => { $action }
+            $($rest)*
+        )
+    };
+    // simple-match: match on simple patterns that use a block without a comma
+    (
+        match $ty:ident.kind() { $($p:pat => $a:expr),* }
+        BtfKind::$name:ident $(($var:ident))? => $action:block
+        $($rest:tt)*
+    ) => {
+        $crate::__btf_type_match!(match $ty.kind() {
+            $($p => $a,)*
+            $crate::btf::BtfKind::$name => {
+                $(let $var = $crate::btf::types::$name::try_from($ty).unwrap();)*
+                $action
+            }
+        }
+             $($rest)*
+        )
+    };
+    // or-pattern: match on one or more variants without capturing a variable and using an
+    //             expression followed by a comma.
+    (
+        match $ty:ident.kind() { $($p:pat => $a:expr),* }
+        $(BtfKind::$name:ident)|+  => $action:expr,
+        $($rest:tt)*
+    ) => {
+        $crate::__btf_type_match!(
+            match $ty.kind() { $($p => $a),* }
+            $(BtfKind::$name)|* => { $action }
+            $($rest)*
+        )
+    };
+    (
+        match $ty:ident.kind() { $($p:pat => $a:expr),* }
+        $(BtfKind::$name:ident)|+  => $action:block
+        $($rest:tt)*
+    ) => {
+        $crate::__btf_type_match!(match $ty.kind() {
+            $($p => $a,)*
+            $($crate::btf::BtfKind::$name)|* => {
+                $action
+            }
+        }
+             $($rest)*
+        )
+    };
+    // default match case
+    //
+    // we only need the expression case here because this case is not followed by a $rest:tt like
+    // the others, which let's us use the $(,)? pattern.
```

Comment:
Well, it could be ;-) Basically we support
```rust
        btf_type_match!(match ty {
            BtfKind::Int => "int",
            _ => "default",
        });
```
but not
```rust
        btf_type_match!(match ty {
            _ => "default",
            BtfKind::Int => "int",
        });
```
which I think is fair and there is no need, just pointing it out.

Reply:
yeah that was on purpose, the second code will always generate a warning and in general be incorrect. Technically something the vanilla rust match supports but it seemed hardly worth supporting :grin: